src/dataframe/dataframe.jl

```diff
@@ -625,58 +623,93 @@ julia> insertcols!(d, 1, b=['a', 'b', 'c'])
 │ 2   │ 'b'  │ 2     │
 │ 3   │ 'c'  │ 3     │
 
-julia> insertcols!(d, 1, :c => [2, 3, 4])
-3×3 DataFrame
-│ Row │ c     │ b    │ a     │
-│     │ Int64 │ Char │ Int64 │
-├─────┼───────┼──────┼───────┤
-│ 1   │ 2     │ 'a'  │ 1     │
-│ 2   │ 3     │ 'b'  │ 2     │
-│ 3   │ 4     │ 'c'  │ 3     │
+julia> insertcols!(d, 2, :c => 2:4, :c => 3:5, makeunique=true)
+3×4 DataFrame
+│ Row │ b    │ c     │ c_1   │ a     │
+│     │ Char │ Int64 │ Int64 │ Int64 │
+├─────┼──────┼───────┼───────┼───────┤
+│ 1   │ 'a'  │ 2     │ 3     │ 1     │
+│ 2   │ 'b'  │ 3     │ 4     │ 2     │
+│ 3   │ 'c'  │ 4     │ 5     │ 3     │
 ```
 """
-function insertcols!(df::DataFrame, col_ind::Int, name_col::Pair{Symbol, <:AbstractVector};
-                     makeunique::Bool=false)
-    name, item = name_col
+function insertcols!(df::DataFrame, col_ind::Int, name_cols::Pair{Symbol,<:Any}...;
+                     makeunique::Bool=false, copycols::Bool=true)
     if !(0 < col_ind <= ncol(df) + 1)
-        throw(ArgumentError("attempt to insert a column to a data frame with $(ncol(df)) columns at index $col_ind"))
-    end
-    if !(size(df, 1) == length(item) || size(df, 2) == 0)
-        throw(DimensionMismatch("length of new column ($(length(item))) must match the number of rows in data frame ($(nrow(df)))"))
+        throw(ArgumentError("attempt to insert a column to a data frame with " *
+                            "$(ncol(df)) columns at index $col_ind"))
     end
 
-    if hasproperty(df, name)
-        if makeunique
-            k = 1
-            while true
-                # we only make sure that new column name is unique
-                # if df originally had duplicates in names we do not fix it
-                nn = Symbol("$(name)_$k")
-                if !hasproperty(df, nn)
-                    name = nn
-                    break
+    for (name, item) in name_cols
+        if !(item isa AbstractArray)
+            # perform implicit broadcasting of an item that is not an array
+            if ncol(df) == 0
+                target_row_count = 1
+                for (_, v) in name_cols
```

Comment:
Maybe do this loop once at the beginning, and perform the dimension check at the same time? Then this loop will be simpler as you will already know the target row count and error handling won't be necessary anymore.

Reply:
done